nixos/modules/services/web-apps/mastodon.nix

```diff
@@ -0,0 +1,333 @@
+{ config, lib, pkgs, ... }:
+
+let
+  tmpDir = "/tmp/mastodon";
+  logDir = "/var/log/mastodon";
+
+  cfg = config.services.mastodon;
+
+  env = {
+    RAILS_ENV = "production";
+    NODE_ENV = "production";
+
+    VAPID_PUBLIC_KEY = cfg.vapidPublicKey;
+    DB_USER = cfg.dbUser;
+    SMTP_LOGIN  = cfg.smtpLogin;
+
+    REDIS_HOST = cfg.redisHost;
+    REDIS_PORT = toString(cfg.redisPort);
+    DB_HOST = cfg.dbHost;
+    DB_PORT = toString(cfg.dbPort);
+    DB_NAME = cfg.dbName;
+    LOCAL_DOMAIN = cfg.localDomain;
+    SMTP_SERVER = cfg.smtpServer;
+    SMTP_PORT = toString(cfg.smtpPort);
+    SMTP_FROM_ADDRESS = cfg.smtpFromAddress;
+    PAPERCLIP_ROOT_PATH = "${cfg.dataDir}/public-system";
+    PAPERCLIP_ROOT_URL = "/system";
+    ES_ENABLED = if (cfg.elasticsearchHost != null) then "true" else "false";
+    ES_HOST = cfg.elasticsearchHost;
+    ES_PORT = toString(cfg.elasticsearchPort);
+  };
+
+in {
+
+  options = {
+    services.mastodon = {
+      enable = lib.mkEnableOption "Mastodon federated social network";
+      configureNginx = lib.mkOption {
+        description = "Use nginx as reverse proxy for mastodon";
+        type = lib.types.bool;
+        default = true;
+      };
+      user = lib.mkOption {
+        description = ''
+          User under which mastodon runs
+          If it is set to "mastodon", a user will be created.
+        '';
+        type = lib.types.str;
+        default = "mastodon";
+      };
+      group = lib.mkOption {
+        description = ''
+          Group under which mastodon runs
+          If it is set to "mastodon", a group will be created.
+        '';
+        type = lib.types.str;
+        default = "mastodon";
+      };
+
+      streamingPort = lib.mkOption {
+        description = "TCP port used by the mastodon-streaming service";
+        type = lib.types.int;
+        default = 55000;
+      };
+      webPort = lib.mkOption {
+        description = "TCP port used by the mastodon-web service";
+        type = lib.types.int;
+        default = 55001;
+      };
+      sidekiqPort = lib.mkOption {
+        description = "TCP port used by the mastodon-sidekiq service";
+        type = lib.types.int;
+        default = 55002;
+      };
+
+      dbUser = lib.mkOption {
+        description = "Postgres login username";
```

Comment:
Is it much work to add a `dbType` option and support mysql as well?

Reply:
I think this is not supported by mastodon.

Reply:
Ah, ok.